include/oneapi/dpl/internal/distributed_ranges_impl/detail/ranges.hpp

```diff
@@ -0,0 +1,272 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_DR_DETAIL_RANGES_HPP
+#define _ONEDPL_DR_DETAIL_RANGES_HPP
+
+#include <any>
+#include <iterator>
+#include <type_traits>
+
+#include "std_ranges_shim.hpp"
+
+namespace oneapi::dpl::experimental::dr
+{
+
+namespace ranges
+{
+
+template <typename>
+inline constexpr bool disable_rank = false;
+
+namespace __detail
+{
+
+template <typename T>
+concept has_rank_method = requires(T t)
+{
+    {
+        t.rank()
+        } -> std::weakly_incrementable;
+};
+
+template <typename R>
+concept has_rank_adl = requires(R& r)
+{
+    {
+        rank_(r)
+        } -> std::weakly_incrementable;
+};
+
+template <typename Iter>
+concept is_remote_iterator_shadow_impl_ =
+    std::forward_iterator<Iter> && has_rank_method<Iter> && !disable_rank<std::remove_cv_t<Iter>>;
+
+struct rank_fn_
+{
+    // Return the rank associated with a remote range.
+    // This is either:
+    // 1) r.rank(), if the remote range has a `rank()` method
+    // OR, if not available,
+    // 2) r.begin().rank(), if iterator is `remote_iterator`
+    template <stdrng::forward_range R>
+    requires((has_rank_method<R> && !disable_rank<std::remove_cv_t<R>>) ||
```

Comment:
You don't need this requires clause in my opinion because you have all if `constexpr` checks for within the function It just pollutes the code. If the intent is to provide better diagnostics what I would do is writing a `__dependent_false`.

So it's 
```cpp
template <typename... Args>
inline constexpr __dependent_false = false;
```
And then, between line 84 and 85 I would put
```cpp
static_assert(__dependent_false<R>)
// or another trick without __dependent_false.
static_assert(sizeof(R) == 0); // which is always false
```
To me it's much simpler than listing all the conditions twice. You have several such places in this PR